### PR TITLE
Add Blankskip to Zipformer+CTC

### PR DESCRIPTION
**Pull request by yfyeung**
Accelerating RNN-T Training and Inference Using CTC guidance
https://arxiv.org/pdf/2210.16481.pdf

#### greedy_search  (epoch 30 avg 13)
| apply blank skip | test-clean | test-other | decoding time(s) (3090) |
| ------------------ | ---------- | ---------- | ----------------------- |
|  decode  | 8.13       | 17.48      | 41.828                  |
| train&decode | 5.74     | 15.46      | 39.717           |
| None | 5.68       | 15.15      | 48.749              |

#### modified_beam_search (beam-size 4, epoch 30 avg 13)
| apply blank skip | test-clean | test-other | decoding time(s) (3090) |
| ------------------ |  ---------- | ---------- | ----------------------- |
| decode | 7.88       | 17.01      | 100.744                 | 
| train&decode  | 5.66       | 15.24      | 92.286          |
| None | 5.64       | 14.93      | 219.133           |

#### fast_beam_search  (beam 20.0, max-contexts 8, max-states 64, epoch 30 avg 13)
| apply blank skip | test-clean | test-other | decoding time(s) (3090) |
| ------------------ |---------- | ---------- | ----------------------- |
|  decode  | 10.15 | 19.42 | 61.614  |
| train&decode  | 7.72 | 17.51 | 65.103  |
| None | 5.66 | 14.90 | 95.128 |

#### greedy_search (epoch 30 avg 13)
| apply blank skip  | test-clean | test-other | decoding time(s) (V100) |
| ------------------ | ---------- | ---------- | ----------------------- |
|  decode  | 5.84       | 15.43      | 124.449                  |
| train&decode  | 5.92       | 15.59      | 112.946        |
| None |5.67       | 15.15      | 135.733            |

#### modified_beam_search (beam-size 4, epoch 30 avg 13)
| apply blank skip  | test-clean | test-other | decoding time(s) (V100) |
| ------------------ |---------- | ---------- | -----------------------  |
|  decode  | 5.78       | 15.24      | 367.705                  |
| train&decode  | 5.86       | 15.31      | 153.612         |
| None | 5.63       | 14.94      | 431.615            |

### CTC branch
| model | test-clean | test-other | decoding time(s) (3090) | Config |
| --------------------- | ---------- | ---------- | ------- | ------ |
| Zipformer+CTC  |   6.28   |   16.76   |  56.284  |  epoch 30 avg 15 |
| + blank skip && scaling Lconv  |   6.24   |   16.97   |  54.181  | epoch 30 avg 10 |

#### greedy_search
| apply blank skip  | test-clean | test-other | 
| ------------------ |---------- | ---------- | 
| train&decode  | 2.29    | 5.28      |
| None  | 2.23    | 5.19      |

#### modified_beam_search (beam-size 4)
| apply blank skip  | test-clean | test-other | 
| ------------------ |---------- | ---------- | 
| train&decode  | 2.25    | 5.21      |
| None | 2.21       |  5.12      |

### CODE REVIEW

**Comment by yfyeung** [comment]
### with Lconv

**epoch 30 avg 13**

| greedy_search         | test-clean | test-other | Decoding time(s) (3090) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 7.39       | 17.86      | 39.593                  |
| both train and decode | 5.87       | 15.59      | 33.119                  |
| without blank skip    | 5.68       | 15.15      | 40.999                  |


| modified_beam_search  | test-clean | test-other | Decoding time(s) (3090) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 7.41       | 17.84      | 196.817                 |
| both train and decode | 5.78       | 15.38      | 78.604                  |
| without blank skip    | 5.64       | 14.93      | 212.757                 |


### without Lconv 

**epoch 30 avg 13**

| greedy_search         | test-clean | test-other | Decoding time(s) (V100) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 5.84       | 15.43      | 124.449                 |
| both train and decode | 5.92       | 15.59      | 112.946                 |
| without blank skip    | 5.68       | 15.15      |                         |


| modified_beam_search  | test-clean | test-other | Decoding time(s) (V100) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 5.78       | 15.24      | 367.705                 |
| both train and decode | 5.86       | 15.31      | 153.612                 |
| without blank skip    | 5.64       | 14.93      |                         |

**Comment by yfyeung** [comment]
### with Lconv

**epoch 30 avg 13**

| greedy_search         | test-clean | test-other | Decoding time(s) (3090) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 7.39       | 17.86      | 39.593                  |
| both train and decode | 5.87       | 15.59      | 33.119                  |
| without blank skip    | 5.68       | 15.15      | 40.999                  |


| modified_beam_search  | test-clean | test-other | Decoding time(s) (3090) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 7.41       | 17.84      | 196.817                 |
| both train and decode | 5.78       | 15.38      | 78.604                  |
| without blank skip    | 5.64       | 14.93      | 212.757                 |


### without Lconv 

**epoch 30 avg 13**

| greedy_search         | test-clean | test-other | Decoding time(s) (V100) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 5.84       | 15.43      | 124.449                 |
| both train and decode | 5.92       | 15.59      | 112.946                 |
| without blank skip    | 5.67       | 15.15      | 135.733                 |


| modified_beam_search  | test-clean | test-other | Decoding time(s) (V100) |
| --------------------- | ---------- | ---------- | ----------------------- |
| only train            | 5.78       | 15.24      | 367.705                 |
| both train and decode | 5.86       | 15.31      | 153.612                 |
| without blank skip    | 5.63       | 14.94      | 431.615                 |

**Comment by kobenaxie** [comment]
> Accelerating RNN-T Training and Inference Using CTC guidance https://arxiv.org/pdf/2210.16481.pdf
> 
> ## Result
> * If not specifically noted, the result is based on epoch 30 avg 13.
> 
> ### with scaling Lconv
> #### greedy_search
> apply blank skip	test-clean	test-other	decoding time(s) (3090)
> decode	8.13	17.48	41.828
> train&decode	5.74	15.46	39.717
> None	5.68	15.15	48.749
> #### modified_beam_search
> apply blank skip	test-clean	test-other	decoding time(s) (3090)
> decode	7.88	17.01	100.744
> train&decode	5.66	15.24	92.286
> None	5.64	14.93	219.133
> #### fast_beam_search
> apply blank skip	test-clean	test-other	decoding time(s) (3090)
> decode	10.15	19.42	61.614
> train&decode	7.72	17.51	65.103
> None	5.66	14.90	95.128
> ### without Lconv
> #### greedy_search
> apply blank skip	test-clean	test-other	decoding time(s) (V100)
> decode	5.84	15.43	124.449
> train&decode	5.92	15.59	112.946
> None	5.67	15.15	135.733
> #### modified_beam_search
> apply blank skip	test-clean	test-other	decoding time(s) (V100)
> decode	5.78	15.24	367.705
> train&decode	5.86	15.31	153.612
> None	5.63	14.94	431.615
> ### CTC branch
> model	test-clean	test-other	decoding time(s) (3090)	Config
> Zipformer+CTC	6.28	16.76	56.284	epoch 30 avg 15
> + blank skip && scaling Lconv	6.24	16.97	54.181	epoch 30 avg 10

Hi @yfyeung , I also trained zipformer with `Decoder Frame Reduction`, and  modified some lines, I find `blank skip` can also accelerate training (7 cards A100).
|model                                |test-clean|test-other|FR ratio |time per epoch|
|-----------------------------|----------|------------|----------|-----------------|
|zipformer                           |2.15        |5.22          |0%         |59 min             |
|zipformer+blank skip         |2.39        |5.50          |66~70%|39 min            |
|zipformer+CTC+blank skip|2.32        |5.32          |66~70%|39 min            |
* FR ratio: `(length - reduced_length) / length`

**Comment by yfyeung** [comment]
> Hi @yfyeung , I also trained zipformer with `Decoder Frame Reduction`, and modified some lines, I find `blank skip` can also accelerate training (7 cards A100).
> 
> model	test-clean	test-other	FR ratio	time per epoch
> zipformer	2.15	5.22	0%	59 min
> zipformer+blank skip	2.39	5.50	66~70%	39 min
> zipformer+CTC+blank skip	2.32	5.32	66~70%	39 min
> * FR ratio: `(length - reduced_length) / length`
Thanks.

**Comment by yfyeung** [comment]
> Hi @yfyeung , I also trained zipformer with `Decoder Frame Reduction`, and modified some lines, I find `blank skip` can also accelerate training (7 cards A100).
> 
> model	test-clean	test-other	FR ratio	time per epoch
> zipformer	2.15	5.22	0%	59 min
> zipformer+blank skip	2.39	5.50	66~70%	39 min
> zipformer+CTC+blank skip	2.32	5.32	66~70%	39 min
> * FR ratio: `(length - reduced_length) / length`

Thanks.

**Comment by yfyeung** [comment]
> Hi @yfyeung , I also trained zipformer with `Decoder Frame Reduction`, and modified some lines, I find `blank skip` can also accelerate training (7 cards A100).
> 
> model	test-clean	test-other	FR ratio	time per epoch
> zipformer	2.15	5.22	0%	59 min
> zipformer+blank skip	2.39	5.50	66\~70%	39 min
> zipformer+CTC+blank skip	2.32	5.32	66\~70%	39 min
> - FR ratio: `(length - reduced_length) / length`

Thanks.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import math
```
Comment (yaozengwei) [inline egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py:1]: Please add the Copyright info.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    ):
+        super().__init__()
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        x_lens: torch.Tensor,
+        ctc_output: torch.Tensor,
+        blank_id: int = 0,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (yaozengwei) [inline egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py:28]: Please add some documents.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/lconv.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+from typing import List, Optional, Tuple, Union
```
Comment (yaozengwei) [inline egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/lconv.py:1]: Please add the Copyright info.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/test_model.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+from train import get_params, get_transducer_model
+
+
+def test_model_1():
+    params = get_params()
+    params.vocab_size = 500
+    params.blank_id = 0
+    params.context_size = 2
+    params.num_encoder_layers = "2,4,3,2,4"
+    #  params.feedforward_dims = "1024,1024,1536,1536,1024"
```
Comment (glynpu) [inline egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/test_model.py:35]: It could be deleted if it's not needed. (Up to you.)

**Review by glynpu — COMMENTED** [review]
+1

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py (hunk 1 above)
Comment (yfyeung) [reply egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py:20]: ok, I have added.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py (hunk 2 above)
Comment (yfyeung) [reply egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/frame_reducer.py:47]: ok, I have added.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/lconv.py (hunk 3 above)
Comment (yfyeung) [reply egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/lconv.py:17]: ok, I have added.

File: egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/test_model.py (hunk 4 above)
Comment (yfyeung) [reply egs/librispeech/ASR/pruned_transducer_stateless7_ctc_bs/test_model.py]: ok, I have deleted.

**Review by yaozengwei — APPROVED** [review]

**Review by csukuangfj — APPROVED** [review]

**Comment by userasrk2** [comment]
@yfyeung Thanks for merging this,
Can you please explain what does `decoding time(s) (3090)` column mean and how should I translate it to RTF?

Thanks,

**Comment by yfyeung** [comment]
> @yfyeung Thanks for merging this, Can you please explain what does `decoding time(s) (3090)` column mean and how should I translate it to RTF?
> 
> Thanks,

This is the time spent executing the script (decode.py, ctc_guild_decode_bs.py, ctc_decode.py), which is a crude way of measuring decode time. Since it contains the time of loading model, we can`t translate it to RFT.

**Comment by Cescfangs** [comment]
@yfyeung I got a question about decoding time of "LibriSpeech 100h/without lconv/modified beam search

apply blank skip | test-clean | test-other | decoding time(s) (V100)
-- | -- | -- | --
decode | 5.78 | 15.24 | 367.705
train&decode | 5.86 | 15.31 | 153.612
None | 5.63 | 14.94 | 431.615

the decoding time of first row(apply in decode mode only) is much larger than second row, while in other exps the decoding time of these two modes are comparable.

**Comment by yfyeung** [comment]
> @yfyeung I got a question about decoding time of "LibriSpeech 100h/without lconv/modified beam search
> 
> apply blank skip	test-clean	test-other	decoding time(s) (V100)
> decode	5.78	15.24	367.705
> train&decode	5.86	15.31	153.612
> None	5.63	14.94	431.615
> the decoding time of first row(apply in decode mode only) is much larger than second row, while in other exps the decoding time of these two modes are comparable.

It seems the decoding time is not a good indicator. I guess when I was decoding, the cpu was busy with other tasks.

**Comment by yaozengwei** [comment]
> @yfyeung I got a question about decoding time of "LibriSpeech 100h/without lconv/modified beam search
> 
> apply blank skip	test-clean	test-other	decoding time(s) (V100)
> decode	5.78	15.24	367.705
> train&decode	5.86	15.31	153.612
> None	5.63	14.94	431.615
> the decoding time of first row(apply in decode mode only) is much larger than second row, while in other exps the decoding time of these two modes are comparable.

Can you rerun that on a empty gpu, just to ensure no disturbance.

**Comment by Cescfangs** [comment]
@yfyeung theoretically the encoder_output len could be less than token len after frame_reducer(it did happen in my exp), which is not allowed in `get_rnnt_logprobs_smoothed`:

https://github.com/k2-fsa/k2/blob/19cd61b2900e0c59c94189dbb80b791c0465713e/k2/python/k2/rnnt_loss.py#L1035-L1040
 
Is there a workaround to prevent this?

**Comment by yfyeung** [comment]
> @yfyeung theoretically the encoder_output len could be less than token len after frame_reducer(it did happen in my exp), which is not allowed in `get_rnnt_logprobs_smoothed`:
> 
> https://github.com/k2-fsa/k2/blob/19cd61b2900e0c59c94189dbb80b791c0465713e/k2/python/k2/rnnt_loss.py#L1035-L1040
> 
> Is there a workaround to prevent this?

Increase the `warmup` in model.py from 2.0 to 3.0 or higher.

**Comment by yfyeung** [comment]
> @yfyeung theoretically the encoder_output len could be less than token len after frame_reducer(it did happen in my exp), which is not allowed in `get_rnnt_logprobs_smoothed`:
> 
> https://github.com/k2-fsa/k2/blob/19cd61b2900e0c59c94189dbb80b791c0465713e/k2/python/k2/rnnt_loss.py#L1035-L1040
> 
> Is there a workaround to prevent this?

@Cescfangs Increase the warmup in model.py from 2.0 to 3.0 or higher.

**Pull request merged**